presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -287,6 +289,38 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
             return new ExchangeNode(node.getId(), node.getType(), node.getScope(), partitioningScheme, sources, inputs);
         }
 
+        private Symbol canonicalizeExchangeNodeOutput(ExchangeNode node, int symbolIndex, Map<List<Symbol>, Symbol> inputsToOutputs)
+        {
+            Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
+            Optional<Symbol> outputSymbolForInputSymbols = getExchangeNodeOutputSymbolForInputSymbols(node, symbolIndex, inputsToOutputs);
+
+            if (outputSymbolForInputSymbols.isPresent()
+                    && !canonicalOutput.equals(outputSymbolForInputSymbols.get())) {
+                map(canonicalOutput, outputSymbolForInputSymbols.get());
+                return outputSymbolForInputSymbols.get();
+            }
+            else {
+                inputsToOutputs.put(getExchangeNodeInputSymbols(node, symbolIndex), canonicalOutput);
+                return canonicalOutput;
+            }
+        }
+
+        private Optional<Symbol> getExchangeNodeOutputSymbolForInputSymbols(ExchangeNode node, int symbolIndex, Map<List<Symbol>, Symbol> inputsToOutputs)
+        {
+            if (node.getInputs().isEmpty()) {
```

Comment:
I didn't see any assertion in ExchangeNode that there has to be at least 1 source.